tools/canmonitor.cpp

```diff
@@ -64,70 +126,30 @@ int main(int argc, char ** argv)
     std::cout << std::endl;
     std::cout << options.help();
     std::cout << std::endl;
-    return -1;
+    return 0;
   }
 
-  // Read CAN data
-  auto ret = ReturnStatuses::OK;
+  ReturnStatuses ret;
 
   ret = kv_can.open(channel_idx, bitrate);
 
   if (ret == ReturnStatuses::OK)
   {
-    uint32_t id;
-    uint8_t msg[8] = {};  // zero-initialize array
-    uint32_t size;
-    bool extended;
-    uint64_t t;
-    uint16_t no_msg_count = 0;
-
-    while (1)
-    {
-      while ((ret = kv_can.read(&id, msg, &size, &extended, &t)) == ReturnStatuses::OK)
-      {
-        // Write the message to stdout
-        auto unix_timestamp_ms = std::chrono::duration_cast<std::chrono::milliseconds>
-          (std::chrono::system_clock::now().time_since_epoch()).count();
-
-        std::cout << "[" << std::dec << unix_timestamp_ms << "] ";
-        std::cout << "ID 0x" << std::hex << std::uppercase << id << ":";
-        std::cout << std::internal << std::setfill('0');
+    ret = kv_can.registerReadCallback(std::function<void()>(can_read));
 
-        for (uint8_t i = 0; i < size; ++i)
-        {
-          std::cout << " " << std::hex << std::setw(2);
-          std::cout << std::uppercase << static_cast<unsigned int>(msg[i]);
-        }
-
-        std::cout << std::endl;
-      }
-
-      if (ret == ReturnStatuses::NO_MESSAGES_RECEIVED)
-      {
-        no_msg_count++;
-
-        if (no_msg_count > 150)
-        {
-          std::cout << "No messages received." << std::endl;
-          no_msg_count = 0;
-        }
-
-        std::this_thread::sleep_for(std::chrono::milliseconds(10));
-      }
-      else
-      {
-        std::cerr << KvaserCanUtils::returnStatusDesc(ret);
-        shutdown();
-        return -1;
-      }
+    if (ret != ReturnStatuses::OK)
+    {
+      std::cerr << "Error registering reader callback: ";
+      shutdown_with_error(ret, -1);
     }
   }
   else
   {
-    std::cerr << KvaserCanUtils::returnStatusDesc(ret) << std::endl;
-    shutdown();
-    return -1;
+    std::cerr << "Error opening Kvaser interface: ";
+    shutdown_with_error(ret, -2);
   }
 
+  std::getchar();
```

Comment:
I agree but I couldn't find a better method. Since the primary function of this tool is to wait for calbacks, this was the best way I could find to keep the application from exiting without using an infinite loop that wakes up every so often. If there is a better method out there, I'm all ears.

Reply:
Since we've got an infinite loop in that callback, maybe it's better to not use a callback at all. Why is this implementation better than just keeping that loop in main?

Reply:
I mean, this one could really go either way. The callback is more efficient but it's assumed that users won't be using this tool to such an extent where that kind of efficiency would matter and, in this particular use-case, it's maybe more of an anti-pattern. The "infinite" loop in the callback isn't actually "infinite" - it breaks when no more messages are available to be read and exits the callback. Do you have an opinion either way?

Reply:
I'm inclined to agree that this might be a misapplication of the callback pattern. Leaving it as a loop in main is probably going to be good enough for the end-user and easier to maintain down the road. If someone opens an issue complaining about performance it might be good to look back into it, but before then it's not worth it IMO.

Reply:
Changed back to polling loop.